File: apps/scala/design-pattern/test/ProxyTest.scala
package designpattern.proxy

class ProxySuite extends munit.FunSuite:

  test("RealBankAccount で預入・引出できる") {
    val account = RealBankAccount("田中")
    account.deposit(1000)
    assertEqualsDouble(account.balance, 1000.0, 0.01)

    assert(account.withdraw(500))
    assertEqualsDouble(account.balance, 500.0, 0.01)
  }

  test("残高不足の引出は false を返す") {
    val account = RealBankAccount("田中")
    account.deposit(100)
    assert(!account.withdraw(200))
  }

  test("ProtectionProxy は認証なしでアクセスを拒否する") {
    val real  = RealBankAccount("田中")
    val proxy = ProtectionProxy(real, "secret123")

    intercept[SecurityException] {
      proxy.balance
    }
  }

  test("ProtectionProxy は認証後にアクセスを許可する") {
    val real  = RealBankAccount("田中")
    val proxy = ProtectionProxy(real, "secret123")

    assert(proxy.authenticate("secret123"))
    proxy.deposit(1000)
    assertEqualsDouble(proxy.balance, 1000.0, 0.01)
  }

  test("ProtectionProxy は誤ったパスワードで認証に失敗する") {
    val real  = RealBankAccount("田中")
    val proxy = ProtectionProxy(real, "secret123")

    assert(!proxy.authenticate("wrong"))
    intercept[SecurityException] {
      proxy.deposit(1000)
    }
  }

  test("VirtualProxy は lazy val で遅延初期化する") {
    val proxy = VirtualProxy("佐藤")
    // 操作するまで realAccount は初期化されない
    proxy.deposit(500)
    assertEqualsDouble(proxy.balance, 500.0, 0.01)
  }

File: apps/csharp/design-pattern/tests/DesignPattern.Tests/Patterns/ProxyTest.cs
using DesignPattern.Patterns.Proxy;

namespace DesignPattern.Tests.Patterns;

public class ProxyTest
{
    [Fact]
    public void RealBankAccount_Deposit_IncreasesBalance()
    {
        var account = new RealBankAccount(100m);
        account.Deposit(50m);

        Assert.Equal(150m, account.Balance);
    }

    [Fact]
    public void RealBankAccount_Withdraw_DecreasesBalance()
    {
        var account = new RealBankAccount(100m);
        account.Withdraw(30m);

        Assert.Equal(70m, account.Balance);
    }

    [Fact]
    public void RealBankAccount_Withdraw_InsufficientFunds()
    {
        var account = new RealBankAccount(100m);
        var result = account.Withdraw(200m);

        Assert.Contains("Insufficient funds", result);
        Assert.Equal(100m, account.Balance);
    }

    [Fact]
    public void ProtectionProxy_DeniesAccessWithoutAuth()
    {
        var real = new RealBankAccount(100m);
        var proxy = new ProtectionProxy(real, "secret123");

        var result = proxy.Deposit(50m);

        Assert.Contains("Access denied", result);
    }

    [Fact]
    public void ProtectionProxy_AllowsAccessAfterAuth()
    {
        var real = new RealBankAccount(100m);
        var proxy = new ProtectionProxy(real, "secret123");
        proxy.Authenticate("secret123");

        var result = proxy.Deposit(50m);

        Assert.Contains("Deposited", result);
        Assert.Equal(150m, proxy.Balance);
    }

    [Fact]
    public void ProtectionProxy_RejectsWrongPassword()
    {
        var real = new RealBankAccount(100m);
        var proxy = new ProtectionProxy(real, "secret123");
        var authResult = proxy.Authenticate("wrong");

        Assert.Contains("failed", authResult);
    }

    [Fact]
    public void VirtualProxy_DoesNotCreateUntilAccessed()
    {
        var proxy = new VirtualProxy(() => new RealBankAccount(100m));

        Assert.False(proxy.IsCreated);
    }

    [Fact]
    public void VirtualProxy_CreatesOnFirstAccess()
    {
        var proxy = new VirtualProxy(() => new RealBankAccount(100m));

        var balance = proxy.Balance;

        Assert.True(proxy.IsCreated);
        Assert.Equal(100m, balance);
        Assert.Single(proxy.Log);
    }
}
